<img src="../Data/images/ZumiHeader.png" width=700>

# Robot Emotions

<font size =3> Zumi has a personality! In this lesson, you will learn how Zumi detects human emotions as well as how to program Zumi’s personality. You will also learn about sound, how it’s measured, and how it corresponds with emotion.  Finally, you will train your Zumi to recognize and react to her favorite color. </font> 



## How do we detect emotion?
<font size =3> Take a look at the images below and see if you can identify each of the emotions. 

<img src="../Data/images/emotions.png" width=700> <br>

How did you determine which emotion was which? There are many features that can be indicators, like the eyes, mouth, eyebrows, and maybe gestures. How do we translate human emotions to a robot?

If you have seen the movie *Cars*, you may know that each of the cars has a personality. How was each car able to express emotions? Was it through movements? Sounds? Eyes? </font> 


### Import libraries
<font size =3> To use personality functions, we need to import the Zumi, screen, and personality libraries. </font> 


In [1]:
from zumi.zumi import Zumi
from zumi.util.camera import Camera
from zumi.util.screen import Screen
from zumi.personality import Personality
from zumi.util.color_classifier import ColorClassifier
import time


zumi = Zumi()
screen = Screen()
personality = Personality(zumi, screen)
camera = Camera()

Starting Zumi 
Pi Zero I2C is available
Verified Pi Zero is the same
Gyroscope previously calibrated
Zumi board detected
Compass detected
OLED Screen detected
Gyroscope & Accelerometer detected


###  Calling personality functions
<font size =3> Here are some functions you can call:

* happy()
     
* celebrate()
       
* angry()
       
* look_around()

* look_around_open()
       
* disoriented_left()
       
* disoriented_right()

* awake()

For example, 
<font face="Courier">personality.happy()</font> will make Zumi wiggle and make a sound!
            
In the cell below, try testing out some of the personality functions to see what they do. </font>

In [ ]:
# Test Personality code here!

## Sounds

<font size =3>Zumi can play sounds to match her emotions! Sound can be measured in frequency and amplitude. 

*   Frequency is the number of pulses or vibrations per second, and is measured in hertz. The higher the frequency, the higher the pitch of the sound is. 
*   Amplitude is how loud or strong the sound is and is measured in decibels. The higher the amplitude, the louder the sound is. 

Video: [Sound: Wavelength, Frequency, and Amplitude](https://www.youtube.com/watch?v=TsQL-sXZOLc)

What does each emotion sounds like? Is happy a low or high frequency? Is angry a low or high amplitude? How does this apply to Zumi?

You can use <font face="Courier">play_note()</font> to play various notes. The first parameter is the note you want to play (anywhere from C2 to B6). The second parameter is optional and denotes the amount of time you want the note to play in milliseconds. The default value is set to 500ms, but you can change that by adding a second parameter like this: <br><br>
<font face="Courier">play_note(Note.GS3, 400)</font>. <br><br>
This plays the note G Sharp below middle C for 400 milliseconds. Try the code below to hear a scale and then compose your own music!</font>


In [ ]:
from zumi.protocol import Note 
zumi.play_note(Note.C4)
zumi.play_note(Note.D4)
zumi.play_note(Note.E4)
zumi.play_note(Note.F4)
zumi.play_note(Note.G4)
zumi.play_note(Note.A4)
zumi.play_note(Note.B4)
zumi.play_note(Note.C5)

<font size =3> Code your own sounds for happy, sad, angry, or excited. Try out different melodies until you find your favorites. </font>

In [ ]:
# Make your melodies here 🎵 

## Screen

<font size =3> Zumi personality also uses the **OLED** (organic LED) screen to display emotions.
There are many different "eyes" Zumi has:

* <font face="Courier"> close_eyes()</font>
* <font face="Courier"> sleepy_eyes()</font>
* <font face="Courier"> sleepy_left()</font>
* <font face="Courier"> sleepy_right()</font>
* <font face="Courier"> blink()</font>
* <font face="Courier"> look_around_open()</font>
* <font face="Courier"> sleeping()</font>
* <font face="Courier"> look_around()</font>     
* <font face="Courier"> glimmer()</font>
* <font face="Courier"> sad()</font>
* <font face="Courier"> happy()</font>
* <font face="Courier"> hello()</font>
* <font face="Courier"> angry()</font>

To use the screen, call the screen class with a function of your choice. Try this: </font>

In [ ]:
screen.sad()

### Draw Text

<font size =3> Aside from drawing Zumi eyes, you can also have Zumi write messages on the screen! Use the <font face="Courier">draw_text()</font> function to write a message like this: </font>

In [ ]:
screen.draw_text("hello!")

<font size =3> If you want to automatically center the text on the screen, call this function instead: </font>

In [ ]:
screen.draw_text_center("hello!")

<font size =3> If you want to write text with numbers, you need to make sure everything is of the <font face="Courier">String</font> data type. </font>

In [ ]:
number = 10
screen.draw_text("ten " + str(number)) # the str() functions turns the number into a string

<font size =3> You can even make Zumi display the time for you! </font>

In [ ]:
for i in range(0,50):
    screen.draw_text_center(time.ctime())
    time.sleep(0.1)

## Eyes on Me
<font size = 3> In this activity, Zumi will keep her eyes on you! You will write code so that Zumi eyes follow your face when it's the middle, left, or right side of her line of sight. If Zumi can't see you, Zumi will be sad!
This activity uses code from Lesson 4.2 Face Detection. Take a look at that lesson if you haven't yet! </font>
    
### Dividing the frame into left, right, and middle
<font size=3> The first step is to figure out how many pixels wide Zumi's camera images are. To do this, we can access the camera object's properties like this:

In [2]:
print(camera.image_width)

160


<font size=3> By default, the pixel width will always be 160. Although you can change this, we are going to use this valaue in our example. Now, divide this image frame into areas that you will consider the left side of the frame, right side of the frame, and the middle. You can divide it equally or not. It's up to you! Just remember that you only have 160 pixels to divide into three parts and you can't have half of a pixel.

## You Smile, I Smile

<font size=3> Now that you have learned to use sounds and screen to give Zumi a personality, have her react to a smile! In a previous lesson, you probably learned about using <font face="Courier">vision.find_face()</font> to search for faces. What do you think the features are for a smile? Are the darker pixels of our mouths angled up or down? In this lesson, you will call a function to check if the face Zumi is detecting is smiling. If she sees a smile, she will be happy. If not, give Zumi some sad eyes.</font>
    
### vision.find_smile()
<font size=3> The function you will call for this activity is <font face="Courier">vision.find_smile()</font>. This function will return the pixel coordinates if Zumi finds a smile, or <font face="Courier">(-1,-1,-1,-1)</font> if she doesn't. For example, if you take a picture and save it in an image, you can check if Zumi found a smile or not: <br> <br>
    
<font face="Courier">
    
(x,y,w,h) = vision.find_smile(image)<br>
if (x,y,w,h) == (-1,-1,-1,-1): <br>
<span style="margin-left: 40px;"># Zumi did not find a smile :(</span> <br>
else: <br>
<span style="margin-left: 40px;"># Zumi sees a smile! :)</span> <br>
    </font> </font>
    
### Code It!
<font size=3> Now write some code that will take a picture with Zumi's camera and look for a smile. Give Zumi some personality and match her emotions with yours! Fill in the code template below:

In [ ]:
camera.start_camera()
try:
    for i in range(1000):
        # Finish the code!

        
finally:
    camera.close()

<hr>

# Extension Activities <br> 

<img src="../Data/images/physics_extension.jpg" width=70 align="left">

###  Frequency <br> 
<font size =3>Use tuners to identify the frequency, wavelength, and amplitude of different Zumi sounds. Which emotion has the highest or lowest frequency, wavelength, and amplitude? </font><br><br>